### PR TITLE
improvement: Add time type matching existing date type

### PR DESCRIPTION
**Pull request by joshprice**
# Contributor checklist

- [x] Documentation changes
- [x] Features include unit/acceptance tests

### CODE REVIEW

**Comment by joshprice** [comment]
This seems to work once I added the time type to the builtin list. However I've run into an issue where I can't seem to set reasonable defaults, on `:time` or `:date` types: 

```elixir
    attribute :day, :date do
      allow_nil? false
      default ~D[2022-04-04]
    end

    attribute :day_end_time, :time do
      allow_nil? false
      default ~T[15:00:00]
    end
```

**Comment by zachdaniel** [comment]
🤔 yeah that is problematic. We should address that default behavior to only support known value types for sure. This is insufficient logic here:

```elixir
  defp default(%{default: value, type: type}, _) do
    case Ash.Type.dump_to_native(type, value) do
      {:ok, value} when is_struct(value) ->
        to_string(value)

      {:ok, value} ->
        inspect(value)

      _ ->
        "nil"
    end
  rescue
    _ ->
      "nil"
  end
```

**Comment by zachdaniel** [comment]
Maybe even a more generic `ToSql`.

File: lib/ash/type/time.ex [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  A builtin type that can be referenced via `:time`
+  """
+  use Ash.Type
+
+  @impl true
+  def storage_type, do: :time
+
+  @impl true
+  def cast_input(value, _) do
```
Comment (zachdaniel) [inline lib/ash/type/time.ex:13]: We should add a `def cast_input(nil, _), do: {:ok, nil}`

Comment (joshprice) [reply lib/ash/type/time.ex:15]: Fixed

**Comment by joshprice** [comment]
Not sure I quite follow yet what's required for the defaults to work correctly. Can we make that a separate issue?

**Comment by joshprice** [comment]
I wanted to write a test for these changes but noticed that many types weren't tested, I've added a rudimentary time happy case test, but let me know if there's any more cases worth including.

**Comment by joshprice** [comment]
I think all doc changes needed are handled by this line https://github.com/ash-project/ash/blob/main/lib/ash/type/type.ex#L51

**Comment by zachdaniel** [comment]
That looks great :)  A lot of the primitive types just pass through to ecto so they don't all have tests.

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

**Pull request merged**